src/java.base/share/classes/jdk/internal/foreign/LayoutRecordMapper.java

```diff
@@ -0,0 +1,641 @@
+                                    // (MemorySegment, Class leafType, Function mapper)
+                                    mh = MethodHandles.insertArguments(mh, 1, byteOffset);
+
+                                    switch (info.elementLayout()) {
+                                        case ValueLayout vl -> {
+                                            // (MemorySegment, Class leafType, Function mapper) ->
+                                            // (MemorySegment, Function mapper)
+                                            mh = MethodHandles.insertArguments(mh, 1, vl.carrier());
+                                            Function<MemorySegment, Object> leafArrayMapper =
+                                            switch (vl) {
+                                                case ValueLayout.OfByte ofByte ->
+                                                        ms -> ms.toArray(ofByte);
+                                                case ValueLayout.OfBoolean ofBoolean ->
+                                                        throw new UnsupportedOperationException("boolean arrays not supported: " + ofBoolean);
+                                                case ValueLayout.OfShort ofShort ->
+                                                         ms -> ms.toArray(ofShort);
+                                                case ValueLayout.OfChar ofChar ->
+                                                        ms -> ms.toArray(ofChar);
+                                                case ValueLayout.OfInt ofInt ->
+                                                        ms -> ms.toArray(ofInt);
+                                                case ValueLayout.OfLong ofLong ->
+                                                        ms -> ms.toArray(ofLong);
+                                                case ValueLayout.OfFloat ofFloat ->
+                                                        ms -> ms.toArray(ofFloat);
+                                                case ValueLayout.OfDouble ofDouble ->
+                                                        ms -> ms.toArray(ofDouble);
+                                                case AddressLayout addressLayout ->
+                                                        ms -> ms.elements(addressLayout)
+                                                                .map(s -> s.get(addressLayout, 0))
+                                                                .toArray(MemorySegment[]::new);
+                                            };
+                                            // (MemorySegment, Function mapper) ->
+                                            // (MemorySegment)
+                                            mh = MethodHandles.insertArguments(mh, 1, leafArrayMapper);
+                                            yield castReturnType(mh, cl.component().getType());
```

Comment:
I'm not sure about this and other casts in this branch of the code. isn't the method handle here returning some T[] ? If so, how can we cast it to the component type of cl? (btw, we already have `cl.componentType` stored in a variable, I think).

Reply:
> I'm not sure about this and other casts in this branch of the code

Never mind, I confused myself between record components and Class::componentType (for the array element type) :-)